doc/specs/#2634 - Broadcast Input/#2634 - Broadcast Input.md

```diff
@@ -0,0 +1,353 @@
+  some (sub)set of the panes in the current tab. So global broadcast mode is
+  the only time a user would need to worry about input being to be sent to
+  an inactive tab.
+* You can have a set of panes to broadcast to in the first tab, then a
+  _separate_ set to broadcast to in a second tab. Broadcasting in one tab
+  does not affect the other.
+
+#### Cons
+* I frankly think the `tab`/`pane` interaction can be a little weird.
+* You can't broadcast to a subset of panes in inactive tabs, in addition to
+  the active tab. All panes you want to broadcast to must be in the active
+  tab.
+* Does creating a new split in a pane that's being broadcast to add that pane to
+  the broadcast set?
+
+#### What would this mean for PR #9222?
+
+The prototype PR [#9222] basically just implemented `{ "action":
+"toggleBroadcastInput", "scope": "tab" }`. We could make `tab` the default
+`scope` if no other one is specified, and then the PR would need basically no
+modifications. Future PRs could add args to the `toggleBroadcastInput` action,
+without breaking users who bind a key to that action now.
+
+### Proposal 2: Broadcast Set
+
+This was the design I had originally came up with before investigating iTerm2
+much closer. This design involves a "broadcast set" of panes. All the panes in
+the broadcast set would also get the `KeySent` and `CharSent` events, in
+addition to the active pane. (The active pane may be a part of the broadcast
+set). If a pane is read-only in the broadcast set, then it won't handle those
+broadcasted events (obviously).
+
+As far as actions, we're looking at something like:
+
+* **A** Only send input to the active pane
+  * Remove all the panes from the broadcast set
+* **B** send input to all panes in all tabs
+  * If all the panes are in the broadcast set, remove them all. Otherwise, add
+    all panes in all tabs to the broadcast set.
+* **C** send input to all panes in the current tab
+  * If all the panes in the current tab are in the broadcast set, remove them
+    from the broadcast set. Otherwise, add all the panes from this tab to the
+    broadcast set.
+* **D** toggle sending input to the current pane
+  * If this pane is in the broadcast set, remove it. Otherwise add it.
+This seems to break down into the following actions:
+
+```json
+{ "action": "toggleBroadcastInput", "scope": "none" },
+{ "action": "toggleBroadcastInput", "scope": "global" },
+{ "action": "toggleBroadcastInput", "scope": "tab" },
+{ "action": "toggleBroadcastInput", "scope": "pane" },
+```
+
+Which would be accompanied by the following internal properties:
+* A global (`TerminalPage`-level)  set of panes to broadcast to.
+
+#### Pros:
+* Mentally, you're either adding panes to the set of panes to broadcast to, or
+  removing them.
+* You can broadcast to panes in multiple tabs, without broadcasting to _all_
+  panes in all tabs.
+
+#### Cons:
+* You can't have a set of panes to broadcast to in the one tab, and a different
+  set in another tab.

```

Comment:
So the way I'm imagining proposal 2 is something like: 
1. in tab 1, you add panes A and B to the broadcast set. Typing in either one goes to both A and B.
2. in tab 1, switch to pane C. Now input goes to A, B and C.
3. in tab 1, switch to pane D. Now input goes to A, B and D.
4. switch to tab 2, pane E. Now input goes to A, B and E.

You can't have like, a set with A & B (in 1), then E & F (in 2). So if someone wants to type to both panes in 1, then both panes in 2, then both panes in 1, they need to keep toggling which panes are in the broadcast set.

Reply:
I think that's fine? Because the act of "enabling broadcast input to a pane" means that you're marking that pane as a target of your input. So switching tabs keeps that state.

Then adding/removing all the panes in a tab with the "scope: tab" just keeps updating that state. It's a more convenient way than adding the panes in one-by-one.